### DISCUSSION TITLE
Deserialize string into tags

### DISCUSSION THREAD

**Discussion by oneingan** [tweag/nickel#864] [Q&A]
I'd like to use an Enum contract with and imported YAML file. But there are not tags in YAML, only strings. Is there a way to deserialize the string into a tag. Or maybe I need to make a new function for it?

**Comment by yannham** [comment]
Hm, that's not an easy one, because there is no reason for the deserializer to choose a tag over a string.

One possibility is to have a contract do that for you. Contract shouldn't modify values in an arbitrary way, but however "normalization" is seen as ok. By normalization, I mean an operation that may chose a uniform representation, or do an automatic conversion, but doesn't change the content of the data. Concretely, this operation must be idempotent, which is just a pedantic word to say that if you apply it once, then applying the transformation again shouldn't do anything (a normalized value won't be changed).

In particular, you can make an extended enum contract that also accepts strings, and if the input is a string, convert it to an enum before checking:

```
nickel> 
let EnumOrString = fun enum_contract label value =>
  let as_enum_tag = if builtin.is_str value then
    string.to_enum value
    else value in
  contract.apply enum_contract label as_enum_tag in

let Schema = {
  foo | EnumOrString [| `foo, `bar, `baz |],
  bar | EnumOrString [| `bar, `babar |],
  ..
} in

# it's hardcoded here, but replace by an import
let imported | Schema = {
  foo = "foo",
  bar = `bar,
  extra = 2
} in

record.values imported
[ `bar, 2, `foo ]

nickel> 
```

You can adapt if you don't care about specifying the list of allowed tags via the `enum_contract` (then just check that the input is either a string or an enum, and return as_enum_tag, or fail otherwise), and probably find a better name for the contract :sweat_smile: 

Otherwise you can indeed just traverse the record in a more traditional way using nested `record.map` and applying the equivalent of `EnumOrString` as a function to values of interest. You probably need to do that in particular if you don't know statically the shape of your imported data in advance.
